ambuda/views/auth.py

```diff
@@ -105,14 +167,76 @@ def sign_out():
     return redirect(url_for(POST_AUTH_ROUTE))
 
 
-@bp.route("/recover", methods=["GET", "POST"])
-def recover():
-    form = RecoverForm()
+@bp.route("/reset-password", methods=["GET", "POST"])
+def get_reset_password_token():
+    """Email the user a password reset link."""
+    form = ResetPasswordForm()
     if form.validate_on_submit():
-        hashed = generate_password_hash(form.password.data)
-    else:
-        hashed = ""
-    return render_template("auth/recover.html", form=form, hashed=hashed)
+        email = form.email.data
+        session = q.get_session()
+        user = session.query(db.User).filter_by(email=email).first()
+        if user:
+            raw_token = _create_reset_token(user.id)
+            mail.send_reset_password_link(
+                username=user.username, email=user.email, raw_token=raw_token
+            )
+            return render_template("auth/reset-password-post.html", email=user.email)
+        else:
+            flash(
+                "Sorry, the email address you provided is not associated with any of our acounts."
+            )
+
+    # Override the default message ("The response parameter is missing.")
+    # for better UX.
+    if form.recaptcha.errors:
+        form.recaptcha.errors = ["Please click the reCAPTCHA box."]
+
+    return render_template("auth/reset-password.html", form=form)
+
+
+@bp.route("/reset-password/<username>/<raw_token>", methods=["GET", "POST"])
+def reset_password_from_token(username, raw_token):
+    """Reset password after the user clicks a reset link."""
+    msg_invalid = "Sorry, this reset password link isn't valid. Please try again."
+
+    user = q.user(username)
```

Comment:
As an extra precaution, we might also want to require the user to enter in their username when they reset their password. This way, if password reset emails are forwarded, or someone requests password resets for multiple accounts, it makes it less likely one would unintentionally reset the wrong user's account -- as well as make it clearer to the user which account's password they are resetting.

Reply:
I'd like to hold off on this since I don't have a good sense of the best UX here. For now, the email with the password reset link contains the corresponding username.